compiler/ast/trees.nim

```diff
@@ -44,61 +44,64 @@ proc sameFloatIgnoreNan(a, b: BiggestFloat): bool {.inline.} =
   ## ignores NaN semantics, but ensures 0.0 == -0.0, see #13730
   cast[uint64](a) == cast[uint64](b) or a == b
 
-proc exprStructuralEquivalent*(a, b: PNode; strictSymEquality=false): bool =
-  if a == b:
-    result = true
-  elif (a != nil) and (b != nil) and (a.kind == b.kind):
-    case a.kind
-    of nkSym:
-      if strictSymEquality:
-        result = a.sym == b.sym
-      else:
-        # don't go nuts here: same symbol as string is enough:
-        result = a.sym.name.id == b.sym.name.id
-    of nkIdent: result = a.ident.id == b.ident.id
-    of nkIntLiterals: result = a.intVal == b.intVal
-    of nkFloatLiterals: result = sameFloatIgnoreNan(a.floatVal, b.floatVal)
-    of nkStrLiterals: result = a.strVal == b.strVal
-    of nkCommentStmt: result = a.comment == b.comment
-    of nkEmpty, nkNilLit, nkType: result = true
-    of nkError:
-      unreachable()
-    of nkWithSons:
-      if a.len == b.len:
-        for i in 0..<a.len:
-          if not exprStructuralEquivalent(a[i], b[i],
-                                          strictSymEquality): return
-        result = true
+template makeTreeEquivalenceProc*(
+  name, relaxedKindCheck, symCheck, floatCheck, typeCheck, commentCheck) {.dirty.} =
+  # Defines a tree equivalence checking procedure.
+  # This skeleton is shared between all recursive
+  # PNode equivalence checks in the compiler code base
+  # It might be possible to unify more of them with each other.
+  proc name(a, b: PNode): bool =
+    result = false
+    if a == b:
+      result = true
+    elif a != nil and b != nil and (a.kind == b.kind or relaxedKindCheck):
+      case a.kind
+      of nkError:           unreachable()
+      of nkEmpty, nkNilLit: result = true
+      of nkSym:             result = symCheck
+      of nkIdent:           result = a.ident.id == b.ident.id
+      of nkIntLiterals:     result = a.intVal == b.intVal
+      of nkFloatLiterals:   result = floatCheck
+        # XXX: Using float equality, even if partially tamed through
+        # sameFloatIgnoreNan, causes inconsistencies due to it lacking
+        # the substition and reflexivity property.
+      of nkStrLiterals:     result = a.strVal == b.strVal
+      of nkType:            result = typeCheck
+      of nkCommentStmt:     result = commentCheck
+      # XXX: nkNimNodeLit should probably always be checked strictly.
+      # Currently this doesn't matter as only nkSym NimNode literals are created
```

Comment:
```suggestion
      #      Currently this doesn't matter as only nkSym NimNode literals are created
```

Align with the above so it's clear that it's part of the same comment.

Also, if that's the case this should be enforced via an assertion.

Reply:
Removed in ea33c54054580cd75880b41b5cfa215e6e170ba6 as it's not actually the case (thanks to @zerbina for pointing that out).